### PR TITLE
Schema version checks

### LINKED ISSUE #2233: Remerge: add sanity checks around changes in schema/version
Follow up from #2226.

E.g. if you load a db with a schema version 0.1.0 and then 0.0.1 (or whatever), we'll accept it and replace the schema. We probably should either return an error, or perhaps open using the newer schema? Not clear, but whatever we decide to do is probably not going to be hard to implement.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-413)
┆Story Points: 3
┆Epic: Remerge production readiness
┆Sprint: SYNC - end 2020-03-13

### PR DESCRIPTION
**Pull request by glasserc**
Fixes #2233.

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - As far as I can test on this machine, this is true.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - This is a change to Remerge internals which has not been mentioned in changelogs so far
  - No breaking changes to Swift or Kotlin binding APIs
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - No new dependencies.

### CODE REVIEW

File: components/remerge/src/engine.rs [hunk 1]
```diff
@@ -398,4 +398,50 @@ mod test {
         assert_eq!(um2["bar"], "test");
         um2.assert_tombstones(vec!["foo", "quux"]);
     }
+
+    #[test]
+    fn test_schema_cant_go_backwards() {
+        const FILENAME: &str = "test_schema_go_backwards.sqlite";
+        let _e: RemergeEngine = RemergeEngine::open(FILENAME, &*SCHEMA).unwrap();
```
Comment (glasserc) [inline components/remerge/src/engine.rs]: I wasn't sure how best to test this kind of thing. I need a persistent SQLite database so I can compare "previous" to "current". This creates sqlite files but they seem to pile up in `components/remerge/` and clutter `git status`. I'm not sure where these should go or how to get them there.

File: components/remerge/src/engine.rs [hunk 2]
```diff
@@ -398,4 +398,50 @@ mod test {
...
+        let backwards_schema: String = json!({
+            "version": "0.1.0",
+            "name": "logins-example",
+            "fields": [],
+        })
+        .to_string();
+        let open_result = RemergeEngine::open(FILENAME, &*backwards_schema);
+        std::fs::remove_file(FILENAME).expect("uh oh, couldn't clean up test");
```
Comment (glasserc) [inline components/remerge/src/engine.rs]: I'm not sure of how best to clean up the SQLite databases when I'm done with them. This gets rid of the actual database but leaves `-shm` and `-wal` files around.

File: components/remerge/src/storage/bootstrap.rs [hunk 3]
```diff
@@ -36,8 +36,27 @@ pub(super) fn load_or_bootstrap(
 
         if native_ver != native.version.to_string() {
             // XXX migrate existing records here!
-            // XXX Ensure we only move native version forward and not backwards!
+            let native_ver = semver::Version::parse(&*native_ver)
+                .expect("previously-written version is no longer semver");
```
Comment (glasserc) [inline components/remerge/src/storage/bootstrap.rs]: This did not seem worth surfacing as an error constructor, but `.expect()` here isn't great. Thoughts?

File: components/remerge/src/storage/bootstrap.rs [hunk 4]
```diff
@@ -36,8 +36,27 @@ pub(super) fn load_or_bootstrap(
...
+                ));
+            }
             meta::put(db, meta::NATIVE_SCHEMA_VERSION, &native.version.to_string())?;
+        } else {
+            let previous_native: String = db.query_row(
+                "SELECT schema_text FROM remerge_schemas WHERE version = ?",
+                rusqlite::params![native_ver],
+                |r| r.get(0),
+            )?;
+            let previous_native = crate::schema::parse_from_string(&*previous_native, false)?;
```
Comment (glasserc) [inline components/remerge/src/storage/bootstrap.rs]: This feels costly but seemed more correct so as not to be confused by whitespace changes.

File: components/remerge/src/engine.rs (hunk 2 above)
Comment (thomcc) [reply components/remerge/src/engine.rs]: RemergeEngine::open_in_memory is a thing, or creating it with the ":memory:" filename (https://www.sqlite.org/inmemorydb.html)

File: components/remerge/src/storage/bootstrap.rs (hunk 4 above)
Comment (thomcc) [reply components/remerge/src/storage/bootstrap.rs]: Only happens at startup so I'm not worried.

File: components/remerge/src/engine.rs (hunk 2 above)
Comment (glasserc) [reply components/remerge/src/engine.rs]: How do I create an in-memory database that persists so I can access it again later?

Comment (linacambridge) [reply components/remerge/src/engine.rs]: I think you can use a URI filename with a `mode=memory` param, like [here](https://www.sqlite.org/inmemorydb.html#sharedmemdb).

Comment (thomcc) [reply components/remerge/src/engine.rs]: Alternatively, if you can't make this work, you could put it inside the `target` directory, which will wipe it out with `cargo clean` and keep it from git status/etc

Be sure to delete it at the start of the test if you do this though.

Comment (glasserc) [reply components/remerge/src/engine.rs]: `mode=memory&cache=shared` worked like a charm, thanks @linacambridge!

**Pull request merged**